### ISSUE TITLE
Serverside validation

### ISSUE THREAD

**Issue by SamMousa** [surveyjs/survey-library#3057]
I am working on a server side docker based verification tool here: https://github.com/HeRAMS-WHO/surveyjs-validator

@andrewtelnov if ever in the future you guys want to adopt it I'd definitely be open to that. As you can see from the code in the repo, it doesn't do anything complicated.

This issue is mainly so when others search for it they'll find this. Feel free to close at any time.

**Comment by andrewtelnov** [comment]
@SamMousa Thank you! We will keep in mind. We are busy with SurveyJS Creator V2 right now. However, we are thinking about creating Server Side functionalities and this can be one of them.

Thank you,
Andrew

**Comment by omasson-perso** [comment]
Hi,

I'll also be interested by server side functionalities.
When the hasErrors function validates data, the return data is not normalized (but you normalize it in the validator).

I'd like to normalize data (a text field with numeric validator should be set to 30 and not '30')

Thanks a lot,

**Comment by andrewtelnov** [comment]
It has been implemented long time ago: https://surveyjs.io/form-library/documentation/data-validation#server-side-validation

**Comment by SamMousa** [comment]
No, that's not what this issue is about. What you link to is server side validation of individual questions. I'm talking about enforcing the built in validation on the server. So when I add a regex rule to the survey json, I reuse the same SurveyJS code and JSON to validate it on the server.

**Comment by andrewtelnov** [comment]
Hello Sam,
That's supported, and it works exactly the way you describe, by reusing the same SurveyJS code and the same JSON on the server.

[survey-core](https://www.npmjs.com/package/survey-core) is a UI-independent library that runs in Node.js. When you instantiate a model from your survey JSON there, all built-in validation defined in that JSON: regex validators, required flags, numeric ranges, expression validators, `isRequired` on matrix columns, and so on — is enforced by the same engine code that runs in the browser:

```js
const { Model } = require("survey-core");

function validateResponse(surveyJson, response) {
  const survey = new Model(surveyJson);
  survey.data = response;
  // Strips values that don't match any question or available choice,
  // and values of questions hidden by visibility logic
  survey.clearIncorrectValues(true);
  const isValid = survey.validate(); // runs all validators from the JSON
  return { isValid, data: survey.data };
}
```

The same code if there is an async expressions in the JSON:
```js
const { Model } = require("survey-core");

function validateResponse(surveyJson, response) {
  const survey = new Model(surveyJson);
  survey.data = response;
  // Strips values that don't match any question or available choice,
  // and values of questions hidden by visibility logic
  survey.clearIncorrectValues(true);
  // Runs all validators from the JSON. Validation can be asynchronous
  // (e.g. expression validators with custom async functions), so the
  // result is delivered via a callback, wrapped here in a Promise:
  return new Promise((resolve) => {
    survey.validate(true, false, (hasErrors) => {
      resolve({ isValid: !hasErrors, data: survey.data });
    });
  });
}

// Usage:
const { isValid, data } = await validateResponse(surveyJson, response);
```



So a regex rule added to the survey JSON is checked on the server with zero duplicated logic, the JSON is the single source of truth.

We also provide this packaged as a deployable backend service with a Docker setup: [surveyjs/surveyjs-json-schema-validator](https://github.com/surveyjs/surveyjs-json-schema-validator). Its `POST /response` endpoint validates a response object against a survey schema using the engine as above, and `POST /schema` additionally validates the survey definition itself (unknown properties, unknown variables in expressions, missing required attributes).

Thank you,
Andrew
